torch/csrc/jit/passes/memory_planning.cpp

```diff
@@ -0,0 +1,273 @@
+  std::vector<int64_t> sizes;
+  auto _sizes = ttp->sizes().concrete_sizes();
+  // TODO: why does this break? answer: in place mutation
+  // also %9 : Long(requires_grad=0, device=cpu) = prim::Constant[value={0}]()
+  if (_sizes.has_value() && _sizes.value().size() > 0 &&
+      _sizes.value()[0] != 0) {
+    sizes = _sizes.value();
+  } else {
+    sizes = std::vector<int64_t>{0};
+  }
+  std::vector<int64_t> strides;
+  auto _strides = ttp->strides().concrete_sizes();
+  if (_strides.has_value() && _strides.value().size() > 0 &&
+      _strides.value()[0] != 0) {
+    strides = _strides.value();
+  } else {
+    strides = at::detail::defaultStrides(sizes);
+  }
+  return std::make_pair(sizes, strides);
+}
+
+Node* insertAllocStorageNode(
+    std::shared_ptr<Graph>& graph,
+    uint64_t total_size) {
+  auto* storage = graph->create(prim::AllocateStorage, 1);
+  storage->i_(attr::total_size, total_size);
+
+  auto deviceType = jit::tensorexpr::pickDeviceType(graph);
+  if (deviceType.has_value()) {
+    storage->i_(attr::device, static_cast<int8_t>(deviceType.value().type()));
+  } else {
+    storage->i_(attr::device, static_cast<int8_t>(at::kCPU));
+  }
+  storage->insertBefore(graph->nodes().front());
+  return storage;
+}
+
+void insertAllocTensorNodes(
+    std::shared_ptr<Graph>& graph,
+    Node* storage,
+    std::unordered_map<LiveRange, Region, live_range_hash> allocations,
+    std::map<LiveRange, const Value*, live_range_start_cmp>
+        manage_range_values) {
+  uint64_t total_size = storage->i(attr::total_size);
+  for (auto& item : manage_range_values) {
+    auto lvr = item.first;
+    auto region = allocations[lvr];
+    auto allocation = item.second;
+
+    // const_cast fishy?
+    auto node = const_cast<Node*>(allocation->node());
+
+    // the way that this node magically *becomes* the out varaint is simply
+    // by add an extra input. this is because op resolution happens
+    // at runtime via the op registry (by matching on the schema).
+    auto* alloc = graph->create(prim::AllocateTensor, 1);
+    node->addInput(alloc->output());
+    GRAPH_DEBUG("inserting allocation op for ", node->getOperator().schema());
+    alloc->insertBefore(node);
+    alloc->addInput(storage->output());
+
+    auto ttp = allocation->type()->expect<c10::TensorType>();
+    std::vector<int64_t> sizes, strides;
+    std::tie(sizes, strides) = getSizesStrides(ttp);
+    TORCH_CHECK(
+        region.offset + region.size <= total_size,
+        "trying to create an allocation that exceeds previously planned memory");
+    alloc->i_(attr::size, region.size);
+    alloc->i_(attr::offset, region.offset);
+    alloc->is_(attr::sizes, sizes);
+    alloc->is_(attr::stride, strides);
+    alloc->i_(attr::device, static_cast<int8_t>(storage->i(attr::device)));
+    alloc->i_(attr::dtype, static_cast<int8_t>(ttp->scalarType().value()));
+  }
+}
+
+bool hasOutVariant(Node* node) {
+  for (const auto& variant : getAllOperatorsFor(node->kind())) {
```

Comment:
This is not the best way of checking because you *could* have an operator overloaded with different schemas where one of the schemas has an out variant but another does not. I think this is what your `aten::cat` alludes to.
Maybe an easier way of doing this is to add a Tensor argument to the end of the operator, check if `maybeSchema` exists, then check that the last argument name is `out`. Then remove the extra tensor argument. For the tensor value you can just construct a prim::Uninitialized Node and set the output type to tensor and use that.